### ISSUE TITLE
Beta: iOS micro:bit app refuses to pair with hex from Beta

### ISSUE THREAD

**Issue by martinwork** [microsoft/pxt-microbit#5913]
Hex files from beta contain version information that the iOS micro:bit app interprets as too old to work. I expect this is a CODAL issue - mentioning here for visibility. I will investigate further.

**Comment by microbit-carlos** [comment]
Thanks for opening the ticket Martin! I didn't realised this was in CODAL, I'll look into this right away.

@abchatra something we need to discuss for the release.

**Comment by microbit-matt-hillsdon** [comment]
Heads up: I don't normally look at the iOS code so this could be wrong!

You can access the beta editor by long pressing the back button inside MakeCode in iOS and enabling the toggle.

Best guess is that it’s reading this:   `NSString *strVersion = [ self jsonTextForKey: irmAT"eVER" inDict: dictHeader];`
 
https://github.com/microbit-foundation/microbit-ios/blob/4c65a90f21699fa4c76a5e6e62e7e690f35a9041/Source/irmHexToBin.m#L1136C55-L1136C59

That seems to be the eVER field in the LZMA’d metadata which uses this value:

```
pxt.appTarget.versions ? pxt.appTarget.versions.target : "",
```

I'm not sure what type of version number that is (perhaps CODAL?).
 
If you test live hex on device -> flash using iOS app with beta your get an “Update” / “Send” / “Cancel” dialog. Send still works. Update appears to do nothing (back to MakeCode).

If you test beta hex on device -> flash using iOS app with beta you also get an error but no dialog just an option to view Details that shows “Failed - old software”.

**Comment by microbit-matt-hillsdon** [comment]
Not convinced its that version any more - lots of versions are checked before showing the UI that reads that version again. Will leave to @martinwork and @carlosperate (who I think suspects CODAL version).

**Comment by martinwork** [comment]
Thanks @microbit-matt-hillsdon @microbit-carlos It refuses to pair with a recently built C++ program too.

**Comment by microbit-carlos** [comment]
@martinwork I can confirm that this is caused by the latest CODAL populating the `CODAL_VERSION`/`DEVICE_DAL_VERSION` macro with the tag version (in this case `"0.2.67"`) instead of the old `"unkonwn"`.

Tested this by using MakeCode beta with the iOS app and setting the  `CODAL_VERSION` macro in the pxt.json file (although sometimes I needed to exit the editor and reopen the project to be able to get MakeCode to recompile using the cloud compiler). And this does work with beta:

```json
{
    "yotta": {
        "config": {
            "CODAL_VERSION": "\"unknown\""
        }
    }
}
```

If I set `CODAL_VERSION` to `"2.0.0"` it does seem to work (although in some cases it says that the MakeCode version is old, offers the option to "update" "send" or "cancel", and using "send" works).

If I set it to `"1.0.0"`, and the micro:bit is running  a hex from live (version "unknown"), then it also shows the "old MakeCode" error, but can click "send" and it flashes. end up with the "Failed - old software" "Old firmware" pairing error.

But once the board has a hex with CODAL_VERSION set to "1.0.0" (via WebUSB flashing or BLE flashing), then I cannot flash anymore and get a "Failed - old software" "Old firmware" error.

Is it possible the app is checking the "firmware version" to determine if it's compatible with the micro:bit board connected?

**Comment by martinwork** [comment]
Yes the checks in the app are for V1. V2 always passed by default because it's version was "unknown", but now fails because the V2 DAL version is earlier than any V1 DAL version. The thresholds are... reject versions <1.4.16 and ask about sending a version < 2.1.0 unless the target is a V1.3. The app needs to be updated to cope with the change, which might take a few days.

**Comment by microbit-carlos** [comment]
Thanks Martin!
I'll update CODAL to still set CODAL_VERSION to `"unknown"` as it will affect all iOS app users.

We should consider this a blocking issue, so I'll reopen this until I have send a PR, and test the latest beta with the iOS app.

@abchatra once I've release a new CODAL tag and it's merged into MakeCode, will somebody from the MakeCode team be able to test the iOS app with MakeCode beta before releasing to live?

Any previous testing with the app should have triggered this issue, so it'd be good if you could also check if you can replicate the problem as it currently stands, to be able to confidently test that it's been fixed.

**Comment by microbit-carlos** [comment]
@abchatra I've created the PR:
- https://github.com/microsoft/pxt-microbit/pull/5916

**Comment by martinwork** [comment]
@microbit-carlos The app seems OK with MakeCode 7.0.38.

**Comment by microbit-carlos** [comment]
Thanks Martin!
I've also tested the Android app with a combination of old hex files, Makecode v6.0.28 and v7.0.38 and all works.

**Comment by microbit-carlos** [comment]
This can be closed as fixed via https://github.com/microsoft/pxt-microbit/pull/5916.
